lib/diff.go

```diff
@@ -117,16 +117,63 @@ func (m *DatasetMethods) Diff(ctx context.Context, p *DiffParams) (*DiffResponse
 			return nil, err
 		}
 	}
+	got, _, err := m.inst.Dispatch(ctx, dispatchMethodName(m, "diff"), p)
+	if res, ok := got.(*DiffResponse); ok {
+		return res, err
+	}
+	return nil, dispatchReturnError(got, err)
+}
 
-	res := &DiffResponse{}
+func schemaDiff(ctx context.Context, left, right *component.BodyComponent) ([]*Delta, *DiffStat, error) {
+	dd := deepdiff.New()
+	if left.Format == ".csv" && right.Format == ".csv" {
+		left, err := terribleHackToGetHeaderRow(left.InferredSchema)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	if m.inst.http != nil {
-		err := m.inst.http.Call(ctx, AEDiff, p, &res)
+		right, err := terribleHackToGetHeaderRow(right.InferredSchema)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-		return res, nil
+
+		return dd.StatDiff(ctx, left, right)
 	}
+	return dd.StatDiff(ctx, left.InferredSchema, right.InferredSchema)
+}
+
+// TODO (b5) - this is terrible. We need better logic error handling for
+// jsonschemas describing CSV data. We're relying too heavily on the schema
+// being well-formed
+func terribleHackToGetHeaderRow(sch map[string]interface{}) ([]string, error) {
+	if itemObj, ok := sch["items"].(map[string]interface{}); ok {
+		if itemArr, ok := itemObj["items"].([]interface{}); ok {
+			titles := make([]string, len(itemArr))
+			for i, f := range itemArr {
+				if field, ok := f.(map[string]interface{}); ok {
+					if title, ok := field["title"].(string); ok {
+						titles[i] = title
+					}
+				}
+			}
+			return titles, nil
+		}
+	}
+	log.Debug("that terrible hack to detect header row & types just failed")
+	return nil, fmt.Errorf("nope")
+}
```

Comment:
remove this & replace all callers with [`tabular.ColumnsFromJSONSchema`](https://github.com/qri-io/dataset/blob/3e9020e68b0e6b789bbcc2b38cc90224b22a1058/tabular/tabular.go#L125)